DOCSP-15338 crud delete (#69)
mongodb/docs-java: source/fundamentals/crud/write-operations/delete.txt

=================

> Delete a Document

Reviewer: Should we mention [findOneAndDelete](https://mongodb.github.io/mongo-java-driver/4.3/apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson,com.mongodb.client.model.FindOneAndDeleteOptions))?

=================

.. default-domain:: mongodb

Overview
--------

> You can remove documents by passing a query filter into the

Reviewer: ```suggestion
You can remove documents by passing a query filter to the
```

:java-docs:`deleteOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteOne(org.bson.conversions.Bson)>` 
or :java-docs:`deleteMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteMany(org.bson.conversions.Bson)>` methods.

The ``deleteOne()`` method deletes a single document. If the query
filter matches more than one document, the method will remove the first
occurrence of a match in the collection. 

The ``deleteMany()`` method deletes all documents that match the query
filter.

Examples
--------

The following examples are about a paint store that sells eight different
colors of paint. The store had their annual online sale resulting in the
following documents in their ``paint_inventory`` collection: 

.. code-block:: json

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 3, "color": "blue", "qty": 0 }
    { "_id": 4, "color": "white", "qty": 0 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 6, "color": "pink", "qty": 0 }
    { "_id": 7, "color": "green", "qty": 0 }
    { "_id": 8, "color": "black", "qty": 8 }

> Using ``deleteMany()``:

Reviewer: ```suggestion
``deleteMany()``:
```

```````````````````````

> All the paint colors in the collection are visible on their site. The
> owner would like to change the site to show colors in stock (``qty`` is
> not ``0``).  

Reviewer: ```suggestion
The paint store website displays all documents in the ``paint_inventory``
collection. To reduce customer confusion, any colors that are out of stock
should be removed.  
```

Just a quick stab at this. It can probably be improved upon.

To remove the colors out of stock, query the ``paint_inventory``
collection where the ``qty`` is ``0`` and pass the query to the

> ``deleteMany()`` method.

Reviewer: "colors out of stock" feels awkward. Perhaps "out of stock items"?

collection: 

.. code-block:: json    

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 8, "color": "black", "qty": 8 }

Using ``deleteOne()``:
``````````````````````

The store is donating the remaining quantity of their purple paint.
This means that the ``qty`` for purple is now ``0`` and we need to
remove purple from the collection. 

To remove purple, query the ``paint_inventory`` collection where the
``color`` is ``"purple"`` and pass the query to the ``deleteOne()``
method. 

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
   :language: java
   :dedent:
   :start-after: begin deleteOneExample
   :end-before: end deleteOneExample

In this scenario, you can also query the ``paint_inventory`` collection
where the ``qty`` is ``8``.  This works because purple is the first
occurrence that matches the query filter. 

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
   :language: java
   :dedent:
   :start-after: begin deleteOneEqExample
   :end-before: end deleteOneEqExample

> 

Reviewer: I think this would be better served as a note in most cases, as it is bonus material that may not be obvious. However, in this case let's remove it since it talks about ordering. Unless there's an index that's being used or a sort is specified, order isn't guaranteed.

=================
Delete a Document
=================

.. default-domain:: mongodb

Overview
--------

> You can remove documents by passing a query filter in the

Reviewer: ```suggestion
You can remove documents by passing a query filter to the
```
I think it's more correct to pass thing _to_ something.

:java-docs:`deleteOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteOne(org.bson.conversions.Bson)>`, 
:java-docs:`deleteMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteMany(org.bson.conversions.Bson)>` 
or :java-docs:`findOneAndDelete() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson)>` methods.

The ``deleteOne()`` method deletes a single document. If the query
filter matches more than one document, the method will remove the first
occurrence of a match in the collection. 

The ``deleteMany()`` method deletes all documents that match the query
filter.

> The ``findOneAndDelete()`` method deletes the first occurrence of a

Reviewer: I think this can be expanded upon?

Like ``deleteOne()``, ``findOneAndDelete()`` deletes the first occurrence of a matching document in a collection. However, the find and delete portions are *atomic*...

match in the collection. 

Examples
--------

The following examples are about a paint store that sells eight different
colors of paint. The store had their annual online sale resulting in the
following documents in their ``paint_inventory`` collection: 

.. code-block:: json

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 3, "color": "blue", "qty": 0 }
    { "_id": 4, "color": "white", "qty": 0 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 6, "color": "pink", "qty": 0 }
    { "_id": 7, "color": "green", "qty": 0 }
    { "_id": 8, "color": "black", "qty": 8 }

``deleteMany()``:
`````````````````

The paint store website displays all documents in the
``paint_inventory`` collection. To reduce customer confusion, the store

> wants to remove the colors that are out of stock (``qty`` is ``0``).

Reviewer: ```suggestion
wants to remove the colors that are out of stock.
```
I don't think this is necessary here, especially since it is detailed immediately in the next paragraph.

To remove the out of stock colors, query the ``paint_inventory``
collection where the ``qty`` is ``0`` and pass the query to the
``deleteMany()`` method.

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
   :language: java
   :dedent:
   :start-after: begin deleteManyExample
   :end-before: end deleteManyExample

The following shows the documents remaining in the ``paint_inventory``
collection: 

.. code-block:: json    

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 8, "color": "black", "qty": 8 }

``deleteOne()``:
````````````````

The store is donating the remaining quantity of their red paint.

> This means that the ``red`` for red is now ``0`` and we need to

Reviewer: ```suggestion
This means that the ``qty`` for red is now ``0`` and we need to
```
Typo?

remove red from the collection. 

To remove red, query the ``paint_inventory`` collection where the
``color`` is ``"red"`` and pass the query to the ``deleteOne()``
method. 

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
   :language: java
   :dedent:
   :start-after: begin deleteOneExample
   :end-before: end deleteOneExample

.. note::

    In this scenario, you can also query all the documents in the
    ``paint_inventory`` collection and pass the query to the
    ``findOneAndDelete()`` method.  This works because red is the first
    occurrence that matches the query filter. 

    .. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
        :language: java
        :dedent:
        :start-after: begin findOneAndDeleteExample

>         :end-before: end findOneAndDeleteExample

Reviewer: In my previous comment I meant we should remove the mention of ordering since it can be a deep subject. Maybe clarifying a bit and using "only" instead of "first"?

   :start-after: begin deleteManyExample
   :end-before: end deleteManyExample

The following shows the documents remaining in the ``paint_inventory``
collection: 

.. code-block:: json    

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 5, "color": "yellow", "qty": 6 }
    { "_id": 8, "color": "black", "qty": 8 }

``deleteOne()``
```````````````

The store is donating the remaining quantity of their yellow paint. This
means that the ``qty`` for yellow is now ``0`` and we need to remove yellow
from the collection. 

To remove yellow, query the ``paint_inventory`` collection where the
``color`` is ``"yellow"`` and pass the query to the ``deleteOne()``
method:

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
   :language: java
   :dedent:
   :start-after: begin deleteOneExample
   :end-before: end deleteOneExample

The following shows the documents remaining in the ``paint_inventory``
collection:

.. code-block:: json   

    { "_id": 1, "color": "red", "qty": 5 }
    { "_id": 2, "color": "purple", "qty": 8 }
    { "_id": 8, "color": "black", "qty": 8 }

``findOneAndDelete()``
``````````````````````

> The store would like to raffle the remaining quantity of a color of
> paint they have the eight of. This means that the ``qty`` of this color
> will be ``0`` and we need to remove this color from the collection.

Reviewer: ```suggestion
The store would like to raffle the remaining quantity of purple
paint, and remove it from the ``paint_inventory`` collection.
```

I think this simplifies it a bit and still gets the message across. Thoughts?

To pick a color, query the ``paint_inventory`` collection where the

> ``qty`` is ``8`` and pass the query to the ``findOneAndDelete()``

Reviewer: ```suggestion
``color`` is ``"purple"`` and pass the query to the ``findOneAndDelete()``
```

method:  

.. literalinclude:: /includes/fundamentals/code-snippets/Delete.java
    :language: java
    :dedent:
    :start-after: begin findOneAndDeleteExample
    :end-before: end findOneAndDeleteExample

> The method returns the document deleted:

Reviewer: ```suggestion
Unlike the other delete methods, ``findOneAndDelete()`` returns
deleted document:
```

.. code-block:: json   

    { "_id": 2, "color": "purple", "qty": 8 }

> This means the store is putting all their purple paint into the raffle.

Reviewer: ```suggestion
```

=================
Delete a Document
=================

.. default-domain:: mongodb

Overview
--------

You can remove documents by passing a query filter to the
:java-docs:`deleteOne() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteOne(org.bson.conversions.Bson)>`, 
:java-docs:`deleteMany() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#deleteMany(org.bson.conversions.Bson)>` 
or :java-docs:`findOneAndDelete() </apidocs/mongodb-driver-sync/com/mongodb/client/MongoCollection.html#findOneAndDelete(org.bson.conversions.Bson)>` methods.

The ``deleteOne()`` method deletes a single document. If the query
filter matches more than one document, the method will remove the first
occurrence of a match in the collection. 

The ``deleteMany()`` method deletes all documents that match the query
filter.

The ``findOneAndDelete()`` method atomically finds and deletes the first
occurrence of a match in the collection.

> If you need to specify a collation or index options to use, you can use
> :java-docs:`DeleteOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/DeleteOptions.html>`
> as a second parameter in ``deleteOne()`` and ``deleteMany()``. For
> ``findOneAndDelete()``, you can also specify a projection, sort, and the
> amount of time the method should run using 
> :java-docs:`FindOneAndDeleteOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>`.
> 

Reviewer: ```suggestion
To specify a collation or hint an index, use
:java-docs:`DeleteOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/DeleteOptions.html>`
as a second parameter to the ``deleteOne()`` and ``deleteMany()`` methods. 

To specify a collation, hint an index, specify sort order, or specify a projection on the returned document, use :java-docs:`FindOneAndDeleteOptions </apidocs/mongodb-driver-core/com/mongodb/client/model/FindOneAndDeleteOptions.html>`
as the second parameter to the ``findOneAndDelete()`` method.
```
